### PR TITLE
Bugfix/parse once

### PR DESCRIPTION
**Pull request by leissa**
# Fixes:

* #99 
* #116
    For now, the `XXX` within `.import XXX` must be a single name in your search paths
* #143 
    There were multiple reads of some files that sometimes caused funny behavior.
* #171
    You can now load a plugin using `.plugin XXX;` in your `.thorin` file.
    
# Changes

* All the bits and pieces which have been passed around manually to make the plugin architecture work are now streamlined and centrally manged in `Driver`.
* File names & paths now use `fs::path` - including `Loc`.
* The c++ sources now use the term `plugin` instead of `dialect`
* more migration from `const Def*` -> `Ref`
* a little bit more of Infer magic
* Wrapper `assert_emplace(container, val)` to assert that an insertion actually took place
* Wrapper `lookup(map, key)` that returns a pointer to value if found or `nullptr` otherwise.

# Remarks

Although we can define now all kind things in our `imports`, I'm reluctant to actually use this feature because right now everything lives in a global namespace. We probably want to have sth smarter.

### CODE REVIEW

**Comment by NeuralCoder3** [comment]
For clarification:

plugin
=> make passes normalizers ... available
import 
=> makes axioms, local definitions, ... in thorin available

Is this correct?

And the relation would be:
plugin -> import
but not import -> plugin
.

**Comment by NeuralCoder3** [comment]
> Although we can define now all kind things in our imports, I'm reluctant to actually use this feature because right now everything lives in a global namespace. We probably want to have sth smarter.

Prefixes like `%name.name` are reserved for axioms (we can not prefix other declarations/definitions), right?

**Comment by NeuralCoder3** [comment]
Is it currently just not necessary that dialects load other dialects using `plugin` or should this be avoided in general?

File: thorin/dump.cpp [hunk 1]
```diff
@@ -382,7 +382,7 @@ void World::dump(std::ostream& os) {
         auto dep    = DepTree(*this);
         auto dumper = Dumper(os, &dep);
 
-        for (const auto& import : imported()) print(os, ".import {};\n", import);
+        for (auto [_, name] : driver().imports()) print(os, ".import {};\n", name);
```
Comment (NeuralCoder3) [inline thorin/dump.cpp:385]: If possible, it would probably be nice to also dump loaded plugins as `.plugin`.

File: cli/main.cpp [hunk 2]
```diff
@@ -45,14 +44,14 @@ int main(int argc, char** argv) {
         auto cli = lyra::cli()
             | lyra::help(show_help)
             | lyra::opt(show_version             )["-v"]["--version"           ]("Display version info and exit.")
-            | lyra::opt(list_dialect_paths       )["-l"]["--list-dialect-paths"]("List search paths in order and exit.")
+            | lyra::opt(list_search_paths        )["-l"]["--list-search-paths" ]("List search paths in order and exit.")
             | lyra::opt(clang,          "clang"  )["-c"]["--clang"             ]("Path to clang executable (default: '" THORIN_WHICH " clang').")
-            | lyra::opt(plugins,        "dialect")["-d"]["--dialect"           ]("Dynamically load dialect [WIP].")
-            | lyra::opt(dialect_paths,  "path"   )["-D"]["--dialect-path"      ]("Path to search dialects in.")
+            | lyra::opt(plugins,        "dialect")["-d"]["--dialect"           ]("Dynamically load plugin.")
```
Comment (NeuralCoder3) [inline cli/main.cpp:49]: If dialects are now called plugins, the option should probably be
`-p` `--plugin`
and
`-P` `--plugin-path`.

**Comment by leissa** [comment]
> > Although we can define now all kind things in our imports, I'm reluctant to actually use this feature because right now everything lives in a global namespace. We probably want to have sth smarter.
> 
> Prefixes like `%name.name` are reserved for axioms (we can not prefix other declarations/definitions), right?

correct. I'd like to change this but this is sth for another PR.

File: cmake/Thorin.cmake [hunk 3]
```diff
@@ -101,7 +101,7 @@ function(add_thorin_dialect)
         OUTPUT ${DIALECT_MD} ${DIALECT_H}
```
[the remark this answers is not in the document]
Comment (leissa) [reply cmake/Thorin.cmake:101]: yes, this is why I'm gradually doing this renaming. I think just renaming files isn't a big deal. Git should be able to cope with it. Only the include section may get messed up in a merge. But again, more PRs are following.

File: cli/main.cpp (hunk 2 above)
Comment (leissa) [reply cli/main.cpp:49]: yes. Also thought about this, but for now I didn't want to touch each and every test case

File: thorin/dump.cpp (hunk 1 above)
Comment (leissa) [reply thorin/dump.cpp:385]: good point. I will do that.

**Pull request merged**